### DISCUSSION TITLE
How do I create my own store

### DISCUSSION THREAD

**Discussion by mtelahun** [maxcountryman/axum-login#28] [Q&A] [answered]
I'm using Sea-ORM for a project of mine and I wanted use axum-login to handle authentication. I'm using postgres as the database so I know I can use the postgres feature. But I was wondering if I can create my own store using Sea-ORM. If so, how would I go about doing that? 

I tried implementing it on my own using memory_store.rs and sql_store.rs as examples but the compiler keeps complaining about AuthUser<Role> not being satisfied for my User struct (even though I believe I've implemented it). Can you give me some pointers?
Thanks.

**Comment by maxcountryman** [comment]
I'm not too familiar with SeaORM but I believe it uses sqlx under the covers.

Can you share an example so I can take a closer look?

**Comment by smorks** [comment, marked as answer]
after struggling a bit to get my own store working, i just recently got it working with Sea-ORM. here's my store, hope it helps you. let me know if you need anything clarified! also, i've been on and off with rust over the years, so the code itself might not be greatest!

```rust
use axum::async_trait;
use axum_login::{secrecy::SecretVec, AuthUser, UserStore};

use ba_core::{sea_orm::DatabaseConnection, Query};

type Result<T = ()> = std::result::Result<T, eyre::Error>;

#[derive(Debug, Clone)]
pub struct TestUser {
    pub id: String,
    pub password: String,
}

impl<Role> AuthUser<Role> for TestUser
where
    Role: PartialOrd + PartialEq + Clone + Send + Sync + 'static,
{
    fn get_id(&self) -> String {
        self.id.clone()
    }

    fn get_password_hash(&self) -> axum_login::secrecy::SecretVec<u8> {
        SecretVec::new(self.password.clone().into())
    }
}

#[derive(Debug, Clone)]
pub struct TestUserStore {
    conn: DatabaseConnection,
}

impl TestUserStore {
    pub fn new(conn: &DatabaseConnection) -> Self {
        Self { conn: conn.clone() }
    }
}

#[async_trait]
impl<Role> UserStore<Role> for TestUserStore
where
    Role: PartialOrd + PartialEq + Clone + Send + Sync + 'static,
{
    type User = TestUser;

    async fn load_user(&self, user_id: &str) -> Result<Option<Self::User>> {
        // my user id is a Vec<u8>, so it's stored base64 encoded
        let id = base64::decode(user_id)?;
        let user = Query::find_user_by_id(&self.conn, &id).await?;
        match user {
            Some(u) => Ok(Some(TestUser {
                id: base64::encode(u.id),
                password: u.password.unwrap(),
            })),
            None => Ok(None),
        }
    }
}
```

**Comment by mtelahun** [comment]
Thank you @smorks ! This is exactly what I needed.
@maxcountryman Thanks for such a great library. It's been very useful.

**Comment by smorks** [comment]
you're welcome @mtelahun, and yes, thanks @maxcountryman for the library. it's working great! axum is fun!

**Comment by jdtron** [comment]
I found this today, as I'm trying to implement a user store for SeaORM, as well.

I tried the above code, with a few tweaks in alignment with changes in the traits.
When using the `Auth` type in my login handler, I get the following error:

![image](https://github.com/maxcountryman/axum-login/assets/58289234/c4323689-1e42-4be5-b969-828e66ebeed7)

Has someone ever seen this and is maybe able to help?

Thanks!

**Reply by jdtron** [reply]
I think I solved this myself.

Make sure the `AuthContext` extractor is the **first** extractor, that your handler requires.

For example
```rust
async fn login(mut auth: Auth, Form(credentials): Form<Credentials>) -> impl IntoResponse {
    // ...
}
```

**Reply by maxcountryman** [reply]
What's important here is that the extractor that consumes the body goes **last**. This is related to how axum works internally.
